File: src/main/java/com/sfh/pokeRogueBot/phase/PhaseProcessor.java
package com.sfh.pokeRogueBot.phase;

import com.sfh.pokeRogueBot.browser.BrowserClient;
import com.sfh.pokeRogueBot.file.FileManager;
import com.sfh.pokeRogueBot.model.enums.GameMode;
import com.sfh.pokeRogueBot.model.exception.NotSupportedException;
import com.sfh.pokeRogueBot.phase.actions.*;
import com.sfh.pokeRogueBot.service.ImageService;
import com.sfh.pokeRogueBot.service.WaitingService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

@Slf4j
@Component
public class PhaseProcessor implements ScreenshotClient {

    private final WaitingService waitingService;
    private final BrowserClient browserClient;
    private final ImageService imageService;
    private final FileManager fileManager;

    public PhaseProcessor(WaitingService waitingService, BrowserClient browserClient, ImageService imageService, FileManager fileManager) {
        this.waitingService = waitingService;
        this.browserClient = browserClient;
        this.imageService = imageService;
        this.fileManager = fileManager;
    }

    /**
     * Handles the given phase by performing the actions in the phase and waits the configured time after the phase.
     */
    public void handlePhase(Phase phase, GameMode gameMode) throws Exception {
        PhaseAction[] actionsToPerform = phase.getActionsForGameMode(gameMode);
        for (PhaseAction action : actionsToPerform) {
            handleAction(action);
        }

        waitingService.waitAfterPhase(phase);
    }

    private void handleAction(PhaseAction action) throws Exception {
        if (action instanceof PressKeyPhaseAction pressKeyPhaseAction) {
            browserClient.pressKey(pressKeyPhaseAction.getKeyToPress());
        } else if (action instanceof WaitPhaseAction) {
            waitingService.waitBriefly();
        } else if (action instanceof WaitForTextRenderPhaseAction) {
            waitingService.waitLonger();
        } else if (action instanceof WaitForStageRenderPhaseAction) {
            waitingService.waitEvenLonger();
        } else if (action instanceof TakeScreenshotPhaseAction) {
            takeScreenshot("phase");
        } else {
            throw new NotSupportedException("Unknown action: " + action.getClass().getSimpleName());
        }
    }

    @Override
    public void takeScreenshot(String prefix) {
        try {
            waitingService.waitEvenLonger();
            fileManager.persistBufferedImage(imageService.takeScreenshot(prefix), prefix);
        } catch (Exception e) {
            log.error("error while taking screenshot", e);
        }
    }

}

File: src/main/kotlin/com/sfh/pokeRogueBot/phase/PhaseProcessor.kt
package com.sfh.pokeRogueBot.phase

import com.sfh.pokeRogueBot.file.FileManager
import com.sfh.pokeRogueBot.model.enums.UiMode
import com.sfh.pokeRogueBot.service.ImageService
import com.sfh.pokeRogueBot.service.WaitingService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class PhaseProcessor(
    private val waitingService: WaitingService,
    private val imageService: ImageService,
    private val fileManager: FileManager
) : ScreenshotClient {

    companion object {
        private val logger = LoggerFactory.getLogger(PhaseProcessor::class.java)
    }

    /**
     * Handles the given phase by performing the actions in the phase and waits the configured time after the phase.
     */
    @Throws(Exception::class)
    fun handlePhase(phase: Phase, uiMode: UiMode) {
        phase.handleUiMode(uiMode)
        waitingService.waitBriefly()
    }


    override fun takeTempScreenshot(prefix: String) {
        try {
            fileManager.saveTempImage(imageService.takeScreenshot(prefix), prefix)
        } catch (e: Exception) {
            logger.error("error while taking temp screenshot: ${e.message}")
        }
    }

    override fun persistScreenshot(prefix: String) {
        try {
            fileManager.persistImage(imageService.takeScreenshot(prefix), prefix)
        } catch (e: Exception) {
            logger.error("error while saving screenshot: ${e.message}")
        }
    }
}